src/GraphQL/Federation/Extensions/FederationGraphTypeExtensions.cs

```diff
@@ -89,4 +89,90 @@ public static void ResolveReference<TSourceType>(this ObjectGraphType<TSourceTyp
+    /// <param name="resolver">The function used to resolve the source representation.</param>
+    public static void ResolveReference<TSourceType, TReturnType>(this IInterfaceGraphType<TReturnType> graphType, Func<IResolveFieldContext, TSourceType, TReturnType?> resolver)
+        where TSourceType : new()
+    {
+        graphType.Metadata[RESOLVER_METADATA] = new FederationResolver<TSourceType, TReturnType>(resolver);
+    }
+
+    /// <summary>
+    /// Configures asynchronous resolution of a reference using a task-based resolver function with a return type.
+    /// </summary>
+    /// <typeparam name="TSourceType">The CLR type of the source representation that the resolver handles.</typeparam>
+    /// <typeparam name="TReturnType">The CLR type of the resolved object returned by the resolver.</typeparam>
+    /// <param name="graphType">The graph type to apply the resolver to.</param>
+    /// <param name="resolver">The asynchronous function used to resolve the source representation.</param>
+    public static void ResolveReference<TSourceType, TReturnType>(this IInterfaceGraphType<TReturnType> graphType, Func<IResolveFieldContext, TSourceType, Task<TReturnType?>> resolver)
+        where TSourceType : new()
+    {
+        graphType.Metadata[RESOLVER_METADATA] = new FederationResolver<TSourceType, TReturnType>(resolver);
+    }
+
+    /// <summary>
+    /// Configures resolution of a reference using a data loader-based resolver function with a return type.
+    /// </summary>
+    /// <typeparam name="TSourceType">The CLR type of the source representation that the resolver handles.</typeparam>
+    /// <typeparam name="TReturnType">The CLR type of the resolved object returned by the resolver.</typeparam>
+    /// <param name="graphType">The graph type to apply the resolver to.</param>
+    /// <param name="resolver">The data loader function used to resolve the source representation.</param>
+    public static void ResolveReference<TSourceType, TReturnType>(this IInterfaceGraphType<TReturnType> graphType, Func<IResolveFieldContext, TSourceType, IDataLoaderResult<TReturnType?>> resolver)
+        where TSourceType : new()
+    {
+        graphType.Metadata[RESOLVER_METADATA] = new FederationResolver<TSourceType, TReturnType>(resolver);
+    }
+
+    /// <summary>
+    /// Configures synchronous resolution of a reference using a resolver function with a return type.
+    /// </summary>
+    /// <typeparam name="TSourceType">The CLR type of the entity type, populated by the representation.</typeparam>
+    /// <param name="graphType">The graph type to apply the resolver to.</param>
+    /// <param name="resolver">The function used to resolve the source representation.</param>
+    public static void ResolveReference<TSourceType>(this IInterfaceGraphType<TSourceType> graphType, Func<IResolveFieldContext, TSourceType, TSourceType?> resolver)
+        where TSourceType : new()
```

Comment:
Here the `: new()` clause is added which will force type inference to break for interfaces, which is necessary because the keys are mapped into `TSourceType` before the resolver returns the object type.

In other words, it wouldn't be able to map `id` into `IMedia.Id` because it can't create an `IMedia`.  So we need to supply a implementing type like `Media` so that it can pass in `Media.Id` so that the resolver can create `Book` which also implements `IMedia`.